### PR TITLE
TIAF: Native Test Opt-In Phase 1

### PR DESCRIPTION
**Pull request by jonawals**
## What does this PR do?

This PR enrolls the Phase 1 native tests into TIAF. All tests in this phase have been additionally opted-in to the native sharding optimization with the `TIAF_sharding` flag (see https://github.com/o3de/o3de/pull/15495). Each test in this phase has been tested for race conditions when run with test sharding, however please see the TIAF sharding PR linked above for caveats and best practices when using test sharding.

# Caveats

The table below demonstrates the absolute and normalized test run times for each test target opted into phase 1. When evaluating the importance of opting in a test target for sharding, you should compare the absolute and normalized values of the Standard Regular, Standard Instrumented and Sharding Instrumented. In this instance, `AzToolsFramework.Tests` takes approx. 22m45s with instrumentation for a standard run compared to only 6m18s without instrumentation. Clearly, even though this test target is a low frequency target for change (i.e. it will not frequently be selected by TIAF to be run) this test run time is prohibitively long for TIAF to be useful as this test target will not only cause large spikes in total test run times but is also long enough to be terminated by the O3DE Jenkins job. These large increases in test run times are due to _OpenCppCoverage_ instrumentation and the actual cause has yet to be determined, but for all intents and purposes it can be assumed to be an intrinsic limitation of this instrumentation. However, with test sharding enabled, this test target takes only 1m19s when instrumented, which is not only faster than a standard run without instrumentation (a happy accident, but not the point of this optimization) but now fast enough to be used in TIAF. Thus, it is of the upmost importance that test targets that are stung by these large run time spikes due to instrumentation must be sharded, and any problems derived from sharding must be fixed at the test level by improving the test hygiene to make all tests in the target truly atomic without any shared state. For more information about this topic, please see the test sharing PR linked above.

| Test                             | Standard  |      |           |      | Sharding  |      |           |      |
|----------------------------------|-----------|------|-----------|------|-----------|------|-----------|------|
|                                  | Regular   |      | Instrum.  |      | Regular   |      | Instrum.  |      |
|                                  | Time (ms) | Norm | Time (ms) | Norm | Time (ms) | Norm | Time (ms) | Norm |
| AzToolsFramework.Tests           | 376.422   | 1    | 1366.962  | 3.63 | 32.956    | 0.09 | 79.502    | 0.21 |
| EMotionFX.Editor.Tests           | 144.199   | 1    | 376.894   | 2.61 | 24.919    | 0.17 | 63.291    | 0.44 |
| EMotionFX.Tests                  | 176.95    | 1    | 456.677   | 2.58 | 9.194     | 0.05 | 40.376    | 0.23 |
| AssetProcessor.Tests             | 152.177   | 1    | 355.193   | 2.33 | 14.207    | 0.09 | 33.228    | 0.22 |
| AzCore.Tests                     | 61.668    | 1    | 211.96    | 3.44 | 10.239    | 0.17 | 25.406    | 0.41 |
| ImageProcessingAtom.Editor.Tests | 36.667    | 1    | 39.816    | 1.09 | 30.736    | 0.84 | 35.018    | 0.96 |
| AzFramework.Tests                | 115.163   | 1    | 562.032   | 4.88 | 21.263    | 0.18 | 36.355    | 0.32 |
| LmbrCentral.Editor.Tests         | 38.313    | 1    | 192.511   | 5.02 | 2.796     | 0.07 | 18.226    | 0.48 |
| PhysX.Tests                      | 29.566    | 1    | 30.754    | 1.04 | 2.569     | 0.09 | 7.664     | 0.26 |
| ScriptCanvasTesting.Editor.Tests | 12.73     | 1    | 18.035    | 1.42 | 3.085     | 0.24 | 36.791    | 2.89 |
| EditorLib.Tests                  | 28.546    | 1    | 103.586   | 3.63 | 2.105     | 0.07 | 34.034    | 1.19 |
| EditorPythonBindings.Tests       | 15.789    | 1    | 21.507    | 1.36 | 1.201     | 0.08 | 3.941     | 0.25 |
| AzManipulatorTestFramework.Tests | 7.736     | 1    | 27.681    | 3.58 | 0.895     | 0.12 | 7.569     | 0.98 |
| SceneProcessing.Editor.Tests     | 9.797     | 1    | 39.254    | 4.01 | 1.546     | 0.16 | 15.771    | 1.61 |
| PrefabBuilder.Tests              | 9.817     | 1    | 11.613    | 1.18 | 0.926     | 0.09 | 9.042     | 0.92 |
| DeltaCataloger.Tests             | 5.728     | 1    | 18.666    | 3.26 | 1.285     | 0.22 | 5.149     | 0.90 |
| AWSCore.Tests                    | 8.28      | 1    | 15.889    | 1.92 | 0.955     | 0.12 | 4.374     | 0.53 |
| Atom_RPI.Tests                   | 9.561     | 1    | 22.515    | 2.35 | 1.079     | 0.11 | 8.47      | 0.89 |

## How was this PR tested?

This PR has been tested locally for test sharding compatibility. This PR will be further tested in AR.

### CODE REVIEW

**Comment by hultonha** [comment]
Also @jonawals is the next PR alluded to ready yet?

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp [hunk 1]
```diff
@@ -35,6 +35,14 @@ namespace UnitTest
             : m_application { AZStd::make_unique<AzFramework::Application>() }
         {}
 
+        // Creates a unique filename for the archive to make the tests atomic and prevent file race conditions when
+		// tests are run concurrently
+        static AZStd::string CreateUniqueArchiveName()
+        {
+            auto uuid = AZ::Uuid::Create();
+            return AZStd::string::format("@usercache@/archivetest.%s.pak", uuid.ToFixedString(false).c_str());
```
Comment (hultonha) [inline Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:43]: suggestion:

```
return AZStd::string::format("@usercache@/archivetest.%s.pak", AZ::Uuid::Create().ToFixedString(false).c_str());
```

**Review by hultonha — APPROVED** [review]
Changes look good, I just have one minor suggestion (not critical) and one question in an earlier comment.

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp (hunk 1 above)
Comment (jonawals) [reply Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:43]: Fixed 👍

**Comment by lumberyard-employee-dm** [comment]
It looks like Validation has failed because a tab has been introduced into the `ArchiveCompressionTests.cpp`

```
VALIDATION FAILURE SUMMARY
VALIDATOR_FAILED: TabsValidator /home/runner/work/o3de/o3de/Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp::TabsValidator FAILED TabsValidator - 1 tabs in this file
    First instance of a tab: 
         +        // Creates a unique filename for the archive to make the tests atomic and prevent file race conditions when
    ---> +\t\t// tests are run concurrently

1 failed, 9 passed in 195.36s
Error: Process completed with exit code 1.
```

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp [hunk 2]
```diff
@@ -35,6 +35,14 @@ namespace UnitTest
             : m_application { AZStd::make_unique<AzFramework::Application>() }
         {}
 
+        // Creates a unique filename for the archive to make the tests atomic and prevent file race conditions when
+		// tests are run concurrently
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:39]: This line contains a tab `\t` on it

**Review by spham-amzn — APPROVED** [review]

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp (hunk 2 above)
Comment (lumberyard-employee-dm) [inline Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:39]: This works for this scenario.

I was thinking a simpler approach with less changes would be to add a `AZ::Test::ScopedAutoTempDirectory m_tempDirectory;` member to this fixture.
Then in the constructor for the `ArchiveCompressionTestFixture`, set the `@usercache@` to that directory
```c++
AZ::IO::FileIOBase* fileIo = AZ::IO::FileIOBase::GetInstance();
fileIo->SetAlias("@usercache@", m_tempDirectory.GetDirectory());
```

**Review by lumberyard-employee-dm — COMMENTED** [review]
What occurs on Linux with these test.
Is the TIAF code branch ignored and regular CTest invocation used?

**Comment by hultonha** [comment]
> Also @jonawals is the next PR alluded to ready yet?

Any update on this PR yet?

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp (hunk 1 above)
Comment (hultonha) [reply Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:43]: Thanks!

File: Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp (hunk 2 above)
Comment (jonawals) [reply Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:39]: Fixed 👍

Comment (jonawals) [reply Code/Framework/AzFramework/Tests/ArchiveCompressionTests.cpp:39]: Great suggestion, fixed 👍

**Comment by jonawals** [comment]
> Also @jonawals is the next PR alluded to ready yet?



> What occurs on Linux with these test?
> Is the TIAF code branch ignored and regular CTest invocation used?

Correct :+1

**Pull request merged**